core/pkg/otel/metrics.go

```diff
@@ -0,0 +1,107 @@
+package otel
+
+import (
+	"context"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.opentelemetry.io/otel/metric/unit"
+	"go.opentelemetry.io/otel/sdk/instrumentation"
+	"go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/aggregation"
+	"time"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric/instrument"
+	semconv "go.opentelemetry.io/otel/semconv/v1.18.0"
+)
+
+type MetricsRecorder struct {
+	httpRequestDurHistogram   instrument.Float64Histogram
+	httpResponseSizeHistogram instrument.Float64Histogram
+	httpRequestsInflight      instrument.Int64UpDownCounter
+	impressions               instrument.Int64Counter
+}
+
+func (r MetricsRecorder) HttpAttributes(svcName, url, method, code string) []attribute.KeyValue {
+	return []attribute.KeyValue{
+		semconv.ServiceNameKey.String(svcName),
+		semconv.HTTPURLKey.String(url),
+		semconv.HTTPMethodKey.String(method),
+		semconv.HTTPStatusCodeKey.String(code),
+	}
+}
+
+func (r MetricsRecorder) HttpRequestDuration(ctx context.Context, duration time.Duration, attrs []attribute.KeyValue) {
+	r.httpRequestDurHistogram.Record(ctx, duration.Seconds(), attrs...)
+}
+
+func (r MetricsRecorder) HttpResponseSize(ctx context.Context, sizeBytes int64, attrs []attribute.KeyValue) {
+	r.httpResponseSizeHistogram.Record(ctx, float64(sizeBytes), attrs...)
+}
+
+func (r MetricsRecorder) InFlightRequestStart(ctx context.Context, attrs []attribute.KeyValue) {
+	r.httpRequestsInflight.Add(ctx, 1, attrs...)
+}
+
+func (r MetricsRecorder) InFlightRequestEnd(ctx context.Context, attrs []attribute.KeyValue) {
+	r.httpRequestsInflight.Add(ctx, -1, attrs...)
+}
+
+func (r MetricsRecorder) Impressions(ctx context.Context, key, variant string) {
+	r.impressions.Add(ctx, 1, []attribute.KeyValue{
+		semconv.FeatureFlagKey(key),
+		semconv.FeatureFlagVariant(variant),
+		semconv.FeatureFlagProviderName("flagd"),
+	}...)
+}
+
+func getDurationView(svcName, viewName string, bucket []float64) metric.View {
+	return metric.NewView(
+		metric.Instrument{
+			// we change aggregation only for instruments with this name and scope
+			Name: viewName,
+			Scope: instrumentation.Scope{
+				Name: svcName,
+			},
+		},
+		metric.Stream{Aggregation: aggregation.ExplicitBucketHistogram{
+			Boundaries: bucket,
+		}},
+	)
+}
+
+func NewOTelRecorder(exporter metric.Reader, serviceName string) *MetricsRecorder {
+	const requestDurationName = "http_request_duration_seconds"
+	const responseSizeName = "http_response_size_bytes"
+
+	// create a metric provider with custom bucket size for histograms
+	provider := metric.NewMeterProvider(
+		metric.WithReader(exporter),
+		metric.WithView(getDurationView(requestDurationName, serviceName, prometheus.DefBuckets)),
```

Comment:
Do we have to use prometheus APIs instead of OTel for these metrics?

Reply:
This is for just keeping the same bucket size as it was before moving to OTel as SDK to report metrics (see https://github.com/open-feature/flagd/pull/419#issuecomment-1440806008)

Reply:
Maybe to expand more on this subject: using OTel allows us to be vendor agnostic and provide metrics in different formats. The "output format" currently is wired inside the [`from_config` file](https://github.com/open-feature/flagd/pull/530/files#diff-60ffee6ef9753f574677d132f3e595feee457c5e52c1b1b88063c616a3c6cdf7R57)